torch/csrc/utils/python_arg_parser.cpp

```diff
@@ -253,17 +253,29 @@ void append_overloaded_arg(std::vector<py::handle> &overloaded_args, PyObject* o
   }
 }
 
-auto FunctionParameter::check(PyObject* obj, std::vector<py::handle> &overloaded_args) -> bool
+static auto check_tensor_or_overload(PyObject* obj, std::vector<py::handle> &overloaded_args) -> bool {
+  if (THPVariable_CheckExact(obj)) {
+    // torch.Tensor instances (not subclasses)
+    return true;
+  }
+
+  if (check_has_torch_function(obj)) {
+    // tensor subclasses and unrelated objects with __torch_function__
+    append_overloaded_arg(overloaded_args, obj);
+    return true;
+  } else if (THPVariable_Check(obj)) {
```

Comment:
This looks semantically different; previously we would test if it was `THPVariable_Check` before `check_has_torch_function`. I don't remember if we discussed this in VC; was there a reason for this change?

Reply:
This was previously breaking the design and dispatching over all args instead of just `Tensor` args. In this PR what Nathan did was essentially move the dispatching to only apply on `Tensor` or `TensorList` operands, which is more in line with NumPy's `__array_function__`.

One other difference that has cropped up is that `__array_function__` doesn't typecheck the remaining arguments if dispatching takes place while this does. I'm not sure whether that's desirable or not.

Reply:
> This was previously breaking the design and dispatching over all args instead of just Tensor args. In this PR what Nathan did was essentially move the dispatching to only apply on Tensor or TensorList operands, which is more in line with NumPy's __array_function__.

This sounds fine.

> One other difference that has cropped up is that __array_function__ doesn't typecheck the remaining arguments if dispatching takes place while this does. I'm not sure whether that's desirable or not.

Hmm, I guess it is not the worst thing in the world, though it is pretty weird that you only typecheck some of the arguments and then not the rest.